types/wlr_xdg_foreign_v1.c

```diff
@@ -0,0 +1,458 @@
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zxdg_exported_v1_interface,
+		&xdg_exported_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void xdg_exported_handle_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static const struct zxdg_exported_v1_interface xdg_exported_impl = {
+	.destroy = xdg_exported_handle_destroy
+};
+
+static void xdg_exporter_handle_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static struct wlr_xdg_exporter_v1 *xdg_exporter_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zxdg_exporter_v1_interface,
+		&xdg_exporter_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void disconnect_imported(struct wlr_xdg_imported_v1 *imported) {
+	if (imported->exported != NULL) {
+		imported->exported = NULL;
+		zxdg_imported_v1_send_destroyed(imported->resource);
+
+		wl_list_remove(&imported->exported_destroyed.link);
+		wl_list_init(&imported->exported_destroyed.link);
+	}
+}
+
+static void xdg_exported_handle_resource_destroy(
+		struct wl_resource *resource) {
+	struct wlr_xdg_exported_v1 *exported =
+		xdg_exported_from_resource(resource);
+
+	wlr_signal_emit_safe(&exported->base.events.destroy, NULL);
+	wlr_xdg_foreign_exported_finish(&exported->base);
+
+	wl_list_remove(&exported->xdg_surface_unmap.link);
+	wl_list_remove(&exported->link);
+	free(exported);
+}
+
+static void handle_xdg_surface_unmap(
+		struct wl_listener *listener, void *data) {
+	struct wlr_xdg_exported_v1 *exported =
+		wl_container_of(listener, exported, xdg_surface_unmap);
+	wl_resource_destroy(exported->resource);
+}
+
+static void xdg_exporter_handle_export(struct wl_client *wl_client,
+		struct wl_resource *client_resource,
+		uint32_t id,
+		struct wl_resource *surface_resource) {
+	struct wlr_xdg_exporter_v1 *exporter =
+		xdg_exporter_from_resource(client_resource);
+	struct wlr_surface *surface = wlr_surface_from_resource(surface_resource);
+
+	if (!verify_is_toplevel(client_resource, surface)) {
+		return;
+	}
+
+	struct wlr_xdg_exported_v1 *exported =
+		calloc(1, sizeof(struct wlr_xdg_exported_v1));
+	if (exported == NULL) {
+		wl_client_post_no_memory(wl_client);
+		return;
+	}
+
+	if (!wlr_xdg_foreign_exported_init(&exported->base, exporter->foreign->registry)) {
+		wl_client_post_no_memory(wl_client);
+		free(exported);
+		return;
+	}
+
+	exported->base.surface = surface;
+	exported->resource = wl_resource_create(wl_client,
+		&zxdg_exported_v1_interface,
+		wl_resource_get_version(exporter->resource),
+		id);
+
+	wl_resource_set_implementation(exported->resource, &xdg_exported_impl,
+			exported, xdg_exported_handle_resource_destroy);
+
+	wl_list_insert(&exporter->exports, &exported->link);
+	wl_list_init(&exported->imports);
+	wl_signal_init(&exported->base.events.destroy);
```

Comment:
This is already done by `wlr_xdg_foreign_exported_init`

Reply:
It isn't done but probably should be :)